lib/eosknowledgeprivate/ekn-media-bin.c

```diff
@@ -0,0 +1,1577 @@
+{
+  if (self->ignore_adjustment_changes)
+    return;
+
+  self->position = gtk_adjustment_get_value (adjustment);
+
+  gst_element_seek_simple (self->play,
+                           GST_FORMAT_TIME,
+                           GST_SEEK_FLAG_ACCURATE |
+                           GST_SEEK_FLAG_TRICKMODE |
+                           GST_SEEK_FLAG_FLUSH,
+                           self->position * GST_SECOND);
+}
+
+static gchar *
+on_progress_scale_format_value (GtkScale    *scale,
+                                gdouble      value,
+                                EknMediaBin *self)
+{
+  /* Pad it with spaces, since CSS padding does not work properly in scale value gadget */
+  return g_strdup_printf (" %s ", format_time (value));
+}
+
+static void
+on_volume_popup_show (GtkWidget *popup, EknMediaBin *self)
+{
+  /* Do not hide controls */
+  ekn_media_bin_revealer_timeout (self, FALSE);
+}
+
+static void
+on_volume_popup_hide (GtkWidget *popup, EknMediaBin *self)
+{
+  ekn_media_bin_revealer_timeout (self, TRUE);
+}
+
+static inline GstElement *
+ekn_media_bin_video_sink_new (GtkWidget *widget)
+{
+  const gchar *vendor, *renderer;
+  GdkGLContext *ctx;
+  GstElement *sink;
+
+  /* Create a GL context for OpenGL API to work */
+  ctx = gdk_window_create_gl_context (gtk_widget_get_window (widget), NULL);
+  gdk_gl_context_make_current (ctx);
+
+  /* Get OpenGL vendor string */
+  vendor   = (gchar *) glGetString (GL_VENDOR);
+  renderer = (gchar *) glGetString (GL_RENDERER);
+
+  if (g_str_equal ("nouveau", vendor) ||
+      (g_str_equal ("VMware, Inc.", vendor) && g_strstr_len (renderer, -1, "llvmpipe") ))
+    {
+      /* Nouvea drivers dont work with gtkglsink because it does not support sharing
+       * a gl context from different threads.
+       */
+      if (g_str_equal ("nouveau", vendor))
+        g_info ("gtkglsink does not work with nouveau drivers, falling back to gtksink\n"
+                "OpenGL  : %s\n"
+                "Vendor  : %s\n"
+                "Renderer: %s",
+                glGetString (GL_VERSION),
+                vendor,
+                renderer);
+
+      /* Fallback to cairo sink */
+      sink = gst_element_factory_make ("gtksink", "EknMediaBinVideoSink");
+    }
+  else
+    {
+      /* Use OpenGL sink */
+      sink = gst_element_factory_make ("gtkglsink", "EknMediaBinGLVideoSink");
+    }
+
+  /* Free GL context */
+  gdk_gl_context_clear_current ();
+  g_object_unref (ctx);
+
+  return sink;
+}
+
+/* This function will make playbin show up the first video frame */
+static inline void
+ekn_media_bin_set_state_ready (EknMediaBin *self)
+{
+  if (self->uri && self->video_sink)
+    {
+      g_object_set (self->play, "uri", self->uri, NULL);
+      gst_element_set_state (self->play, GST_STATE_PAUSED);
+    }
+}
+
+static void
+ekn_media_bin_fullscreen_apply (EknMediaBin *self, gboolean fullscreen)
+{
+  if (fullscreen)
+    {
+#if USE_TMP_IMAGE
```

Comment:
What information do you still need to decide whether to use this code path or not?